rts/Lua/LuaSyncedRead.cpp

```diff
@@ -8592,6 +8596,80 @@ int LuaSyncedRead::GetUnitScriptNames(lua_State* L)
 	return 1;
 }
 
+static int TraceRayGroundImpl(lua_State *const L, const float3 &pos, const float3 &dir, const float maxLen, const bool testWater)
+{
+	const float rayLength = CGround::LineGroundWaterCol(pos, dir, maxLen, testWater, CLuaHandle::GetHandleSynced(L));
+	if (rayLength == -1.0f)
+		return 0;
+
+	const auto collisionSpot = pos + dir * rayLength; // FIXME: would be nice if the CGround:: functions returned this so we wouldn't have to recalculate
+
+	lua_pushnumber(L, rayLength);
+	lua_pushnumber(L, collisionSpot.x);
+	lua_pushnumber(L, collisionSpot.y);
+	lua_pushnumber(L, collisionSpot.z);
+	return 4;
+}
+
+/*** Checks for a ground collision in given direction
+ *
+ * @function Spring.TraceRayGroundInDirection
+ *
+ * Checks if there is surface (ground, optionally water) towards a vector
+ * and returns the distance to the closest hit and its position, if any.
+ *
+ * @number posX
+ * @number posY
+ * @number posZ
+ * @number dirX
+ * @number dirY
+ * @number dirZ
+ * @bool[opt=true] testWater
+ * @treturn number rayLength
+ * @treturn number posX
+ * @treturn number posY
+ * @treturn number posZ
+ */
+int LuaSyncedRead::TraceRayGroundInDirection(lua_State* L)
+{
+	const float3 pos(luaL_checkfloat(L, 1), luaL_checkfloat(L, 2), luaL_checkfloat(L, 3));
+	const auto dir = float3(luaL_checkfloat(L, 4), luaL_checkfloat(L, 5), luaL_checkfloat(L, 6)).Normalize();
+	const float maxLen = luaL_optfloat(L, 7, 999999.f);
```

Comment:
maybe default to FarPlane*1.4 as most of the code is doing, but not really an issue.

Reply:
I think defaulting to camera's FarPlane*1.4 is done for traces that have the camera as the origin, but here you can have an arbitrary point as the origin.